lib/semantics/check-arithmeticif.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "attr.h"
+#include "check-arithmeticif.h"
+#include "scope.h"
+#include "semantics.h"
+#include "symbol.h"
+#include "tools.h"
+#include "type.h"
+#include "../evaluate/traversal.h"
+#include "../parser/message.h"
+#include "../parser/parse-tree.h"
+
+namespace Fortran::semantics {
+
+class ArithmeticIfStmtContext {
+public:
+  ArithmeticIfStmtContext(SemanticsContext &context)
+    : messages_{context.messages()} {}
+
+  bool operator==(const ArithmeticIfStmtContext &x) const { return this == &x; }
+  bool operator!=(const ArithmeticIfStmtContext &x) const { return this != &x; }
+
+  // Arithmetic IF statements have been removed from Fortran 2018.
+  // The constraints and requirements here refer to the 2008 spec.
+  void Check(const parser::ArithmeticIfStmt &arithmeticIfStmt) {
+    // R853 Check for a scalar-numeric-expr
+    // C849 that shall not be of type complex.
+    auto &expr{std::get<parser::Expr>(arithmeticIfStmt.t)};
+    if (expr.typedExpr->v.Rank() > 0) {
+      messages_.Say(expr.source,
+          "ARITHMETIC IF statement must have a scalar numeric expression"_err_en_US);
```

Comment:
This is largely a matter of opinion, but I think it's better to make it clear that the problem is with the "scalar" part, not the "numeric" part. E.g. "ARITHMETIC IF expression must be scalar" or "Array not allowed for ARITHMETIC IF expression".

Reply:
I like your message wording better.  Thanks!